server/swap/swap.go

```diff
@@ -288,130 +357,491 @@ type Swapper struct {
 	bTimeout time.Duration
 	// latencyQ is a queue for coin waiters to deal with network latency.
 	latencyQ *wait.TickerQueue
+
+	gracePeriod int32
+
+	handlerMtx sync.RWMutex
+	stop       bool
+
+	liveWaitersMtx sync.Mutex
+	liveWaiters    map[waiterKey]*handlerArgs
+
+	liveAckersMtx sync.Mutex
+	liveAckers    map[uint64]*msgAckers // keyed by server generated message ID
+}
+
+// Config is the swapper configuration settings. A Config instance is the only
+// argument to the Swapper constructor.
+type Config struct {
+	// DataDir is the folder where swap state is stored.
+	DataDir string
+	// State is the state to restore in construction.
+	State *State
+	// AllowPartialRestore indicates if it is acceptable to load only part of
+	// the State if the Swapper's asset configuration lacks assets required to
+	// load all elements of State.
+	AllowPartialRestore bool
+	// Assets is a map to all the asset information, including the asset backends,
+	// used by this Swapper.
+	Assets map[uint32]*LockableAsset
+	// Mgr is the AuthManager for client messaging and authentication.
+	AuthManager AuthManager
+	// A database backend.
+	Storage Storage
+	// BroadcastTimeout is how long the Swapper will wait for expected swap
+	// transactions following new blocks.
+	BroadcastTimeout time.Duration
+}
+
+func (s *Swapper) restoreState(state *State, allowPartial bool) error {
+	// State binary version check should be done when State is loaded.
+	// if state.Version != stateBinaryVersion {
+	// 	return fmt.Errorf("Cannot load swap state with version %d, expected %d",
+	// 		state.Version, stateBinaryVersion)
+	// }
+
+	// Check that the assets required by State are included
+	missingAssets := make(map[uint32]bool)
+	for _, id := range state.Assets {
+		if s.coins[id] == nil {
+			if !allowPartial {
+				return fmt.Errorf("Unable to find backend for asset %d in restore state.", id)
+			}
+			log.Warnf("Unable to find backend for asset %d in restore state.", id)
+			missingAssets[id] = true
+		}
+	}
+
+	// Load the matchTrackers, calling the Contract and Input asset.Backend
+	// methods as needed.
+
+	translateSwapStatus := func(ss *swapStatus, ssd *swapStatusData) error {
+		if swapCoin := ssd.ContractCoinOut; len(swapCoin) > 0 {
+			assetID := ssd.SwapAsset
+			swap, err := s.coins[assetID].Backend.Contract(swapCoin, ssd.ContractScript)
+			if err != nil {
+				return fmt.Errorf("unable to find swap out coin %x for asset %d: %v", swapCoin, assetID, err)
+			}
+			ss.swap = swap
+			ss.swapTime = encode.UnixTimeMilli(ssd.SwapTime)
+		}
+
+		if ssd.SwapConfirmTime != 0 {
+			ss.swapConfirmed = encode.UnixTimeMilli(ssd.SwapConfirmTime)
+		}
+
+		if redeemCoin := ssd.RedeemCoinIn; len(redeemCoin) > 0 {
+			assetID := ssd.RedeemAsset
+			redeem, err := s.coins[assetID].Backend.Input(redeemCoin)
```

Comment:
The idea was to skip spendsCoin check (note the `contractID` arg in `Redemption(redemptionID, contractID []byte)`), but that's not really an issue since `contractID` will necessarily be available of the redeem coin is.  Will try this route.  I think I had a different set of considerations in an earlier draft, so this is probably fine.

Reply:
Yeah, Redemption works fine even if the spendsCoin check probably is not needed in swap state restore.  I'm pretty sure I created the Input method when I was pursing some (fantastically complicated) GobDecoders.... turned out that it was just far easier to copy a bit of extra data.  Anyway, `Input`, `Output` and `UnspendContract` methods of the `asset.Backend` interface are now removed.